Add the neuron JSON Schema

liamstaras opened the pull request:
Thought this was the best way to proceed - we can generate a form directly from it using something like [https://rjsf-team.github.io/react-jsonschema-form/](https://rjsf-team.github.io/react-jsonschema-form/). Does the schema look about right for the required parameters? (I haven't added REGEX yet... but it's not particularly essential for now)

liamstaras committed:
Add the neuron JSON Schema

tnowotny commented on neuron.schema.json:
```diff
@@ -0,0 +1,148 @@
+                    },
+                    "type": {
+                        "type": "string",
+                        "enum": ["scalar","vector"]
```
I *think* these can be all scalar types, including "scalar" but also explicit "float", "double", "int" (and other C integer types). I am not aware of "vector" a s possible variable type here.

liamstaras replied:
whoops, just assumed that without thinking - I'll correct in new commit

neworderofjamie replied:
yeah, all c types (including uint8_t, uint16_t, uint32_t, uint64_t etc) and scalar are what is generally supported here

liamstaras replied:
fair enough - I'll remove validation for now and we could add regex if needed at a later date...

tnowotny commented on neuron.schema.json:
```diff
@@ -0,0 +1,148 @@
+                "additionalProperties": false
+            }
+        },
+        "additionalParameters": {
```
I am not sure what would go here (but could just be me ...)?

liamstaras replied:
that was for the variable known as _extra_global_params_ in Python, and thought it was a more logical format - but could change if having different names if confusing

tnowotny commented on neuron.schema.json:
```diff
@@ -0,0 +1,148 @@
+                "additionalProperties": false
+            }
+        },
+        "additionalVariables": {
```
Same question - what additional variables?

liamstaras replied:
_additional_input_vars_ from Python - not sure of C++ equivalent

neworderofjamie replied:
So these last two things are called "additional input variables" and "extra global parameters" give or take different naming conventions in both Python and C++ so I think they should be called the same here. The input bit is key as additional input variables are used for adding additional (potentially non-linearly combined) inputs to neurons (don't worry @liamstaras , that was more to jog @tnowotny 's memory than for you :smile)

neworderofjamie replied:
The C++ additional input variable syntax is described [here](http://genn-team.github.io/genn/documentation/4/html/de/ded/sectNeuronModels.html#neuron_additional_input)

liamstaras replied:
fine, I'll rename to additionalInputVariables and extraGlobalParameters :wink:

tnowotny replied:
sounds good - and yes, @neworderofjamie , I do now recall vaguely what this was about! Thanks for the pointers.

tnowotny commented on neuron.schema.json:
```diff
@@ -0,0 +1,148 @@
+                    },
+                    "type": {
+                        "type": "string",
+                        "enum": ["scalar","vector"]
```
See comment about variable types above.

liamstaras replied:
sure, will be fixed!

tnowotny commented on neuron.schema.json:
```diff
@@ -0,0 +1,148 @@
+            "description": "The code for the simulation function of the neuron",
+            "type": "array",
+            "items": {
+                "type": "string"
```
Do we want an array of strings for separate code lines or do we want one string (containing line breaks \n for readability)?

liamstaras replied:
I _thought_ an array would be better (_seems_ to be more standard too!) as we can actually have multiple lines in JSON that was (as far as I know using /n just leads to one long line in the source) but either will do... it's all programmatic anyway

neworderofjamie replied:
A (very) quick google suggests that this is a standard approach as it looks slightly prettier than having lots of \ns everywhere

tnowotny commented on neuron.schema.json:
```diff
@@ -0,0 +1,148 @@
+            "type": "boolean",
+            "default": false
+        },
+        "customBody": {
```
I suppose some sort of catch-all? But what is the use case?

liamstaras replied:
again, a variable in the Python class - not really sure of its function though

neworderofjamie replied:
it is an option in Python but I don't think it should be (the Python interface is still a little rough around the edges) and definitely shouldn't be exposed here

liamstaras replied:
sure, will be removed.

tnowotny reviewed (commented):
Overall looks good but I had some questions - also would like to hear Jamie's opinion on those.

liamstaras commented:
All very valid points... no surprises really. My issue is that I'm still not very familiar with the workings of GeNN; I'll have a good read of the documentation. Also not quite sure of the C++/Python differences in implementation, and I guess some of the properties are redundant for that. Can't quite understand the need additional variables/parameters in the Py version though.

tnowotny commented:
Jamie will bring more clarity - some of the additional stuff  was introduced by him while I was too busy to do a proper code review. So I have a number of blind spots where I am less informed than the  users ...

liamstaras committed:
Update neuron.schema.json

Fix issues detailed in https://github.com/GenGeNN/GenGeNN-resources/pull/1#pullrequestreview-409368596

liamstaras commented:
All fixed up in new commit...

tnowotny reviewed (approved):
As far as I can tell this looks good now.
